usr/lib/cca_stdll/cca_specific.c

```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+    case CKA_EC_POINT: typestr = "CKA_EC_POINT"; break;
+    case CKA_ALWAYS_AUTHENTICATE: typestr = "CKA_ALWAYS_AUTHENTICATE"; break;
+    case CKA_IBM_OPAQUE: typestr = "CKA_IBM_OPAQUE"; break;
+    default:
+        sprintf(tbuf, "unknown type 0x%08lx", a->type);
+        typestr = tbuf;
+        break;
+    }
+
+    switch (a->ulValueLen) {
+    case 0:
+        TRACE_DEBUG("  %s: len=0 pValue=%p\n", typestr, a->pValue);
+        break;
+    case 1:
+        TRACE_DEBUG("  %s: len=%lu value=0x%02hhx\n",
+                    typestr, a->ulValueLen, *((uint8_t *)(a->pValue)));
+        break;
+    case 2:
+        TRACE_DEBUG("  %s: len=%lu value=0x%04hx\n",
+                    typestr, a->ulValueLen, *((uint16_t *)(a->pValue)));
+        break;
+    case 4:
+        TRACE_DEBUG("  %s: len=%lu value=0x%08x\n",
+                    typestr, a->ulValueLen, *((uint32_t *)(a->pValue)));
+        break;
+    case 8:
+        TRACE_DEBUG("  %s: len=%lu value=0x%016lx\n",
+                    typestr, a->ulValueLen, *((uint64_t *)(a->pValue)));
+        break;
+    default:
+        TRACE_DEBUG("  %s: len=%lu value:\n", typestr, a->ulValueLen);
+        dump_mem("  ", a->pValue, a->ulValueLen);
+        break;
+    }
+}
+
+/* Debug function: dump list of attribues from a template via TRACE_DEBUG */
+static void dump_template(TEMPLATE *tmpl)
+{
+    DL_NODE *node = NULL;
+    CK_ATTRIBUTE *a = NULL;
+
+    node = tmpl->attribute_list;
+    while (node) {
+        a = (CK_ATTRIBUTE *) node->data;
+        dump_attr(a);
+        node = node->next;
+    }
+}
+#endif
+
+/* cca token type enum, used with analyse_cca_key_token() */
+enum cca_token_type {
+    sec_des_data_key,
+    sec_aes_data_key,
+    sec_aes_cipher_key,
+    sec_hmac_key,
+    sec_rsa_priv_key,
+    sec_rsa_publ_key,
+    sec_ecc_priv_key,
+    sec_ecc_publ_key
+};
+
+/*
+ * Helper function: Analyse given CCA token.
+ * returns TRUE and keytype and keybitsize if token is known and seems
+ * to be valid (only basic checks are done), otherwise FALSE.
+ */
+static CK_BBOOL analyse_cca_key_token(const CK_BYTE *t, CK_ULONG tlen,
+                                      enum cca_token_type *keytype,
+                                      unsigned int *keybitsize)
+{
+    if (t[0] == 0x01 && (t[4] == 0x00 || t[4] == 0x01)) {
+        /* internal secure cca des data key with exact 64 bytes */
+        if (tlen != 64) {
+            TRACE_DEVEL("CCA DES token has invalid token size %lu != 64\n", tlen);
+            return FALSE;
+        }
+        *keytype = sec_des_data_key;
+        if (t[4] == 0x00)
+            *keybitsize = 8 * 8;
+        else if (t[59] == 0x10)
+            *keybitsize = 16 * 8;
+        else if (t[59] == 0x20)
+            *keybitsize = 24 * 8;
+        else {
+            TRACE_DEVEL("CCA DES data key token has invalid/unknown keysize 0x%02x\n", (int)t[59]);
+            return FALSE;
+        }
+        return TRUE;
+    }
+
+    if (t[0] == 0x01 && t[4] == 0x04) {
```

Comment:
2 levels of switch()-es would be easier to read maybe:
```
switch (t[0]) {
case 0x01:
    switch (t[4]) {
    case xxx:
    }
}
```

Reply:
well, no
the idea of the whole function is to group it based on keytypes. Mixing of ifs and switches and more levels may give the function a 1us performance boost but the human readability is much more worse. At least this is my understanding.